### PR TITLE
[R-package] only warn about early stopping and DART boosting being incompatible if early stopping was requested

### LINKED ISSUE #6612: [R-package] [BUG] Using `lgb.train()` with `booster = "dart"` gives warning regardless of `early_stopping_rounds` value
Hi,

I found this (trivial) bug - well, I assume it is a bug so feel free to close if it is not the case. Regardless of the set values of `early_stopping_rounds` the `lgb.tran()`-function throws a warning message when using the `dart`-booster.

 Below is a minimal working example with three cases displaying the behavior, and the source of the (possible) bug. In all cases the warning occurs, and I have only changed the value of `early_stopping_rounds`! 

## MWE

``` r
# load library
library(lightgbm)
```

``` r
# define a simple dataset
DT <- list(
  x = as.matrix(
    mtcars[,
           grep(
             pattern = "mpg",
             value = TRUE,
             x = colnames(mtcars),
             invert = TRUE
           )
    ]

  ),
  y = mtcars$mpg
)
```

``` r
# run with early_stopping_rounds
# set to 0L
lightgbm::lgb.train(
  params    = list(
    objective = "regression",
    boosting  = "dart"
  ),
  data      = lightgbm::lgb.Dataset(
    data  = DT$x,
    label = DT$y
  ),
  verbose = -1,
  early_stopping_rounds = 0L
)
#> Warning in lightgbm::lgb.train(params = list(objective = "regression", boosting
#> = "dart"), : Early stopping is not available in 'dart' mode.
#> LightGBM Model (1 tree)
#> Objective: regression
#> Fitted to dataset with 10 columns
```

``` r
# run with early_stopping_rounds
# set to 1
lightgbm::lgb.train(
  params    = list(
    objective = "regression",
    boosting  = "dart"
  ),
  data      = lightgbm::lgb.Dataset(
    data  = DT$x,
    label = DT$y
  ),
  verbose = -1,
  early_stopping_rounds = 1
)
#> Warning in lightgbm::lgb.train(params = list(objective = "regression", boosting
#> = "dart"), : Early stopping is not available in 'dart' mode.
#> LightGBM Model (1 tree)
#> Objective: regression
#> Fitted to dataset with 10 columns
```

``` r
# run with early_stopping_rounds
# set to NULL
lightgbm::lgb.train(
  params    = list(
    objective = "regression",
    boosting  = "dart"
  ),
  data      = lightgbm::lgb.Dataset(
    data  = DT$x,
    label = DT$y
  ),
  verbose = -1,
  early_stopping_rounds = NULL
)
#> Warning in lightgbm::lgb.train(params = list(objective = "regression", boosting
#> = "dart"), : Early stopping is not available in 'dart' mode.
#> LightGBM Model (1 tree)
#> Objective: regression
#> Fitted to dataset with 10 columns
```

<sup>Created on 2024-08-15 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.4.1 (2024-06-14)
#>  os       Zorin OS 17.1
#>  system   x86_64, linux-gnu
#>  ui       X11
#>  language en_US:en
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       Europe/Copenhagen
#>  date     2024-08-15
#>  pandoc   3.1.11 @ /usr/lib/rstudio/resources/app/bin/quarto/bin/tools/x86_64/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version date (UTC) lib source
#>  cli           3.6.3   2024-06-21 [1] CRAN (R 4.4.1)
#>  data.table    1.15.4  2024-03-30 [1] CRAN (R 4.4.0)
#>  digest        0.6.36  2024-06-23 [1] CRAN (R 4.4.1)
#>  evaluate      0.24.0  2024-06-10 [1] CRAN (R 4.4.0)
#>  fastmap       1.2.0   2024-05-15 [1] CRAN (R 4.4.0)
#>  fs            1.6.4   2024-04-25 [1] CRAN (R 4.4.0)
#>  glue          1.7.0   2024-01-09 [1] CRAN (R 4.4.0)
#>  htmltools     0.5.8.1 2024-04-04 [1] CRAN (R 4.4.0)
#>  jsonlite      1.8.8   2023-12-04 [1] CRAN (R 4.4.0)
#>  knitr         1.47    2024-05-29 [1] CRAN (R 4.4.0)
#>  lattice       0.22-6  2024-03-20 [4] CRAN (R 4.4.0)
#>  lifecycle     1.0.4   2023-11-07 [1] CRAN (R 4.4.0)
#>  lightgbm    * 4.5.0   2024-07-26 [1] CRAN (R 4.4.1)
#>  magrittr      2.0.3   2022-03-30 [1] CRAN (R 4.4.0)
#>  Matrix        1.7-0   2024-04-26 [4] CRAN (R 4.4.0)
#>  purrr         1.0.2   2023-08-10 [1] CRAN (R 4.4.0)
#>  R.cache       0.16.0  2022-07-21 [1] CRAN (R 4.4.0)
#>  R.methodsS3   1.8.2   2022-06-13 [1] CRAN (R 4.4.0)
#>  R.oo          1.26.0  2024-01-24 [1] CRAN (R 4.4.0)
#>  R.utils       2.12.3  2023-11-18 [1] CRAN (R 4.4.0)
#>  R6            2.5.1   2021-08-19 [1] CRAN (R 4.4.0)
#>  reprex        2.1.0   2024-01-11 [3] CRAN (R 4.4.0)
#>  rlang         1.1.4   2024-06-04 [1] CRAN (R 4.4.0)
#>  rmarkdown     2.27    2024-05-17 [1] CRAN (R 4.4.0)
#>  rstudioapi    0.16.0  2024-03-24 [3] CRAN (R 4.4.0)
#>  sessioninfo   1.2.2   2021-12-06 [1] CRAN (R 4.4.0)
#>  styler        1.10.3  2024-04-07 [1] CRAN (R 4.4.0)
#>  vctrs         0.6.5   2023-12-01 [1] CRAN (R 4.4.0)
#>  withr         3.0.0   2024-01-16 [1] CRAN (R 4.4.0)
#>  xfun          0.45    2024-06-16 [1] CRAN (R 4.4.0)
#>  yaml          2.3.8   2023-12-11 [1] CRAN (R 4.4.0)
#> 
#>  [1] /home/serkan/R/x86_64-pc-linux-gnu-library/4.4
#>  [2] /usr/local/lib/R/site-library
#>  [3] /usr/lib/R/site-library
#>  [4] /usr/lib/R/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>

## Source of bug

In the code below the `waning()` will be triggered regardless of the `early_stopping_rounds` value. 

https://github.com/microsoft/LightGBM/blob/602528315c6d5e4ae1e8f5de106cf56cf5869b16/R-package/R/lgb.train.R#L260-L271

Best,

Serkan

### PR DESCRIPTION
**Pull request by serkor1**
Fixes #6612

* Fix: The associated functions have been fixed by adding a nested if-statement that only triggers if the dart-booster is specified with early_stopping_rounds != NULL.

* Unittests: The unit-test in test_parameters.R have been rewritten in the following way

	* The lightgbm()-test have had a expect_no_warning()-clause added if the model is correctly specified.

	* A new unittest have been created that checks wether lgb.cv() throws a warning errorneously.

This commit closes https://github.com/microsoft/LightGBM/issues/6612. :wrench:

### CODE REVIEW

**Comment by serkor1** [comment]
@microsoft-github-policy-service agree

File: R-package/R/lgb.cv.R
Comment (jameslamb) [inline R-package/R/lgb.cv.R:300]:
```suggestion
    if (using_early_stopping){
      warning("Early stopping is not available in 'dart' mode.")
    }
```

Please use braces here. Please try running the linter to see other style recommendations for this repo.

```shell
Rscript .ci/lint-r-code.R R-package/
```

File: R-package/tests/testthat/test_parameters.R
Comment (jameslamb) [inline R-package/tests/testthat/test_parameters.R:94]:
```suggestion
test_that("training should warn if you use 'dart' boosting and requesting early stopping, specified with 'boosting' or aliases", {
```

I think this would be a more accurate description of what this tests.

File: R-package/tests/testthat/test_parameters.R
Comment (jameslamb) [inline R-package/tests/testthat/test_parameters.R:106]:
```suggestion
    # warning: early stopping requested
```

Please consolidate this to 1 line.

File: R-package/tests/testthat/test_parameters.R
Comment (jameslamb) [inline R-package/tests/testthat/test_parameters.R:121]:
```suggestion
    # no warning: early stopping not requested
```

Please consolidate this to one line, and let's drop the link to the original issue. Anyone wanting to understand why this is the way it is can use the git blame to trace back to the pull request and issue (https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#viewing-the-line-by-line-revision-history-for-a-file).

File: R-package/tests/testthat/test_parameters.R
Comment (jameslamb) [inline R-package/tests/testthat/test_parameters.R]:
```suggestion
        , verbose = -1L
```

There's an extra comma here. I suspect this means you didn't run these tests locally before submitting this. Please do that.

```shell
sh build-cran-package.sh --no-build-vignettes
R CMD INSTALL ./lightgbm*.tar.gz
cd R-package/tests
Rscript testthat.R
```

File: R-package/tests/testthat/test_parameters.R
Comment (jameslamb) [inline R-package/tests/testthat/test_parameters.R:138]: Please remove this `desc = ` and `code =`, and make the test match the style (including indentation and using comma-first) of all the other tests in this file.

File: R-package/tests/testthat/test_parameters.R
Comment (jameslamb) [inline R-package/tests/testthat/test_parameters.R:142]:
```suggestion
```

Customizing `learning_rate` is not necessary to test this behavior. Please remove this. And please remove it in the other `lightgbm()` test you modified. I know it was there already (not added by your PR), but seeing it here made me realize that it's unnecessary.

File: R-package/tests/testthat/test_parameters.R
Comment (jameslamb) [inline R-package/tests/testthat/test_parameters.R:156]:
```suggestion
          data = lgb.Dataset(
            data  = train$data
            , label = train$label
          )
```

please use comma-first as all the other R code in this repo does

File: R-package/tests/testthat/test_parameters.R
Comment (jameslamb) [inline R-package/tests/testthat/test_parameters.R:180]:
```suggestion
      })
```

Please consolidate these.

File: R-package/tests/testthat/test_parameters.R
Comment (jameslamb) [inline R-package/tests/testthat/test_parameters.R:173]:
```suggestion
          data = lgb.Dataset(
            data  = train$data
            , label = train$label
          )
```

File: R-package/tests/testthat/test_parameters.R
Comment (jameslamb) [inline R-package/tests/testthat/test_parameters.R:150]:
```suggestion
      # warning: early stopping requested
```

File: R-package/tests/testthat/test_parameters.R
Comment (jameslamb) [inline R-package/tests/testthat/test_parameters.R:167]:
```suggestion
      # no warning: early stopping not requested
```

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks so much!

I've left some suggestions.

* I've changed the title of this.
  - PR titles become changelog notes here, see for example https://github.com/microsoft/LightGBM/releases/tag/v4.5.0
  - so the title should describe the change in a way that answers the question "what changed about LightGBM?", not "which files did this touch?"
* I've added "fixes #6612" to the description
  - see https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
* I've left some suggestions as comments directly on the diff. Please apply them directly in the browser, by clicking "Add to batch" and then commiting all of them.
  - see https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes

### Other notes

If you push additional commits, please don't force-push. All commits here are merged into 1 when a pull request is merged, so force-pushing is unnecessary.

In the future when you contribute here (or in any open source project on GitHub), don't use the target branch (in this case `master`) on your fork. That makes it more difficult for maintainers to collaborate with you and makes it hard to keep your fork up to date with the upstream repo as it changes. Next time, create a separate branch on your fork, e.g. `git checkout -b fix/r-dart-warning`.

File: R-package/tests/testthat/test_parameters.R
Comment (serkor1) [reply R-package/tests/testthat/test_parameters.R]: Hi @jameslamb,

I didn't run the tests locally, this is true. For what it's worth, I tried building the package with the changes I have made with the code you provided - but it will not build it with the additions that I have made. So I was relying on the `workflows` to do the checking.

And on an honest sidenote, I was under the impression that draft `PR`s were only visible to me.

Comment (jameslamb) [reply R-package/tests/testthat/test_parameters.R]: > I was under the impression that draft PRs were only visible to me.

Ah totally understandable! But no, they're visible to everyone on the project.

If you'd like me to stop reviewing until you've taken this out of draft and marked it ready for review, please let me know. This is not urgent, so I'd be happy to back off and give you some space to test things. That's how we usually work in this repo (draft PRs are not reviewed), but I have seen some contributors who open a draft PR and expect a review, so I assumed that was what you wanted here.

Comment (serkor1) [reply R-package/tests/testthat/test_parameters.R]: Allow me to mark it ready for review! :smile: Man, I am learning so much from this simple task. Thank you for your patience!

**Comment by serkor1** [comment]
Hi @jameslamb,

It took a while to finalize this `PR` - for some reason I couldn't get the testing environment up and running locally. In any case, I have used `expect_silent()` which works on Windows and other OS's too - so that was a good idea!

The main difference from your initial review is that I have shortened the description on the unittests as {lintr} gave me an error on the line length. This can be seen in 2a866e89483a2719a43cda16ced07aeaf5cb5171.

I thought this would be an easy and fast fix, so I didn't consider creating a new branch. I am sorry for this - I will never do stuff on main anymore, I promise :crossed_fingers: 

And thank you so much for all your feedback on this PR.

Best,

**Review by jameslamb — APPROVED** [review]
My only other suggestions after reviewing again were minor style things that are all personal preference, so I just pushed a commit directly changing those. We'll merge this once CI passes.

Thanks so much for the help! We appreciate the report and you being willing to come contribute. We'd be happy to have you come back and contribute again any time 😊

**Pull request merged**
